client/session.go

```diff
@@ -563,3 +595,936 @@ func (s *session) Close() error {
+		client, err := peer.GetConnectionPool().NextClient()
+		if err != nil {
+			return err
+		}
+
+		tctx, _ := thrift.NewContext(s.streamBlocksBatchTimeout)
+		result, err = client.FetchBlocks(tctx, req)
+		return err
+	}); err != nil {
+		for i := range batch {
+			s.streamBlocksReattemptFromPeers(batch[i].blocks, enqueueCh)
+		}
+		return
+	}
+
+	// Parse and act on result
+	for i := range result.Elements {
+		if i >= len(batch) {
+			s.log.Errorf("stream blocks response from peer %v returned more IDs than expected", peer.Host().String())
+			break
+		}
+
+		id := result.Elements[i].ID
+		if id != batch[i].id {
+			s.streamBlocksReattemptFromPeers(batch[i].blocks, enqueueCh)
+			s.log.WithFields(
+				xlog.NewLogField("expectedID", batch[i].id),
+				xlog.NewLogField("actualID", id),
+				xlog.NewLogField("indexID", i),
+			).Errorf("stream blocks response from peer %v returned mismatched ID", peer.Host().String())
+			continue
+		}
+
+		for j := range result.Elements[i].Blocks {
+			if j >= len(batch[i].blocks) {
+				s.log.Errorf("stream blocks response from peer %v returned more blocks than expected", peer.Host().String())
+				break
+			}
+
+			block := result.Elements[i].Blocks[j]
+			if block.Start != batch[i].blocks[j].start.UnixNano() {
+				failed := []blockMetadata{batch[i].blocks[j]}
+				s.streamBlocksReattemptFromPeers(failed, enqueueCh)
+				s.log.WithFields(
+					xlog.NewLogField("id", id),
+					xlog.NewLogField("expectedStart", batch[i].blocks[j].start.UnixNano()),
+					xlog.NewLogField("actualStart", block.Start),
+					xlog.NewLogField("indexID", i),
+					xlog.NewLogField("indexBlock", j),
+				).Errorf("stream blocks response from peer %v returned mismatched ID", peer.Host().String())
+				continue
+			}
+
+			if block.Err != nil {
+				failed := []blockMetadata{batch[i].blocks[j]}
+				s.streamBlocksReattemptFromPeers(failed, enqueueCh)
+				s.log.WithFields(
+					xlog.NewLogField("id", id),
+					xlog.NewLogField("expectedStart", batch[i].blocks[j].start.UnixNano()),
+					xlog.NewLogField("actualStart", block.Start),
+					xlog.NewLogField("indexID", i),
+					xlog.NewLogField("indexBlock", j),
+				).Errorf("stream blocks response from peer %v returned mismatched block start", peer.Host().String())
+				continue
+			}
+
+			blocksResult.addBlockFromPeer(id, block)
+		}
+	}
+}
+
+func (s *session) streamBlocksReattemptFromPeers(
+	blocks []blockMetadata,
+	enqueueCh *enqueueChannel,
+) {
+	// Must do this asynchronously or else could get into a deadlock scenario
+	// where cannot enqueue into the reattempt channel because no more work is
+	// getting done because new attempts are blocked on existing attempts completing
+	// and existing attempts are trying to enqueue into a full reattempt channel
+	enqueue := enqueueCh.enqueueDelayed()
+	s.streamBlocksReattemptWorkers.Go(func() {
+		for i := range blocks {
+			// Reconstruct peers metadata for reattempt
+			peersLen := len(blocks[i].reattempt.peersMetadata)
```

Comment:
Kind of a long line, will combine though and add a new line and indent the value assignment rhs.